### ISSUE TITLE
Check conformance with "OGC Timeseries Profile of O&M candidate standard"

### ISSUE THREAD

**Issue by neothemachine** [Reading-eScience-Centre/coveragejson#28]
There's a draft at http://www.opengeospatial.org/standards/requests/137 called "OGC Timeseries Profile of Observations and Measurements candidate standard". Since CoverageJSON supports basic time series, it would be good to check which concepts from the OGC draft are missing in CoverageJSON and how/if they should be added. By doing that it may also define the focus of CoverageJSON more clearly, for example, whether it should retain full metadata or if it is "just" an exchange format from server to web client for visualization purposes etc, keeping only the absolutely necessary elements like domain, range, parameters.

**Comment by neothemachine** [comment]
TP = Timeseries Profile draft

Comparison:
- TP has time-value-pair and domain-range encoding. We focus on domain-range here only.
- TP focuses on single parameters only, which can be continuous or categorical. We allow multiple parameters.
- TP supports per-point metadata: quality, nilReason, comment, relatedObservation, qualifier, processing, source, censoredReason, accuracy, interpolationType, aggregationDuration. Some of those are not supported in CoverageJSON, see next section.
- Point metadata in TP (in domain-range encoding) are encoded as separate "annotation coverage" where the domain consists of the times to which the metadata applies. Compatible with CoverageJSON, although a bit weird since matching is done at coordinate instead of index level.
- TP supports per-timeseries metadata: temporalExtent, baseTime, spacing, status, sampledMedium, intendedObservationSpacing, parameter, cumulative, accumulationIntervalLength, accumulationAnchorTime, startAnchorPoint, endAnchorPoint, maxGapPeriod. These are a mix of actual metadata (temporalExtent, status, sampledMedium etc.) but also just elements that describe the (optionally regular) domain (baseTime, spacing, startAnchorPoint, endAnchorPoint).
- TP supports compact definitions of regular timeseries domains by baseTime and spacing. CoverageJSON requires listing all coordinates for simplicity.
- TP has aggregationDuration as per-point metadata. In CoverageJSON this probably corresponds more generically to the bounds. Similarly, startAnchorPoint and endAnchorPoint just define the left bound of the first and the right bound of the last coordinate.
- TP supports cumulative timeseries, optionally with an accumulationIntervalLength that defines the length of each such interval, starting from an initial accumulationAnchorTime. For plotting it may be useful to indicate the accumulation periods. However, this seems quite domain-specific and out-of-scope for CoverageJSON, rather it should be included as generic metadata in some way.
- TP supports per-point interpolationType (per point because it might switch within a series...) which seems to combine multiple aspects: how/if data values may be interpolated, whether the values are direct or derived (e.g. via statistical methods), which statistical method was used (average vs maximum vs total within last period). CoverageJSON does not cover such metadata, however it seems important enough for other coverage types to look into this further. It would be relevant for plotting as well as processing, however the danger is that it may get too complex or domain-specific.
- TP includes an ObservationProcess that should describe how the data was created. Should be part of general CoverageJSON metadata.
- TP defines collections similar to CoverageJSON collections, and includes some specific metadata like generationDate or temporalExtent.

Blocking things missing in CoverageJSON:
- Free-form text (comment) as per-point metadata. Currently, the range is numeric only, with mappings for categories. Free-form text is thus not supported. This would mean allowing strings as range values. Similarly, relatedObservation as per-point metadata would be a URL which is a subtype of free-form text.
- aggregationDuration as per-point metadata would imply supporting a "duration" type as range value. This could simply be numeric where the parameter unit is "seconds", "hours" etc.
- quality as per-point metadata is specified as the "SWE Common ‘Quality’ union that allows a qualifier to be specified using a Quantity, Quantity Range, Category or Text type". This flexibility is problematic as we force any given parameter to be of a single type only.
- interpolationType covers topics like statistical methods which are worth exploring in general.
- Informative metadata like intendedObservationSpacing, accumulationIntervalLength, or ObservationProcess should have a proper place in a CoverageJSON document, possibly under a separate "metadata" member to allow for easy enumeration of those.

In summary, I would say that much of this is just a matter of where to put the metadata, this would get us maybe to 80% alignment. Which of the other issues is most important to fix?

**Comment by jonblower** [comment]
The first thing that springs to my mind is enabling free-form text or quality info as per-point metadata. This would be a very nice thing to demonstrate in clients and would fit with the CHARMe annotation stuff (somehow). Would it be a big problem to allow strings as range values?

The other metadata is important but a bit less "impactful" if you see what I mean. (However, WP10 and CEH might disagree!)

**Comment by neothemachine** [comment]
I don't think adding support for strings would be problematic. By default, we could say a range is made of numeric values. And if you want strings, then you have to add "dataType": "string" in the range object. If those strings have a special meaning, e.g. they are URIs, then that would have to be put as metadata in the parameter object.

**Comment by jonblower** [comment]
Would it make sense to use "dataType": "uri" if we want clients to interpret the strings as URIs?

**Comment by neothemachine** [comment]
Nope, remember we wanted to keep a strict boundary between range and parameter. Any semantics above pure primitive data value encoding goes into the parameter. (like category integer -> url mappings)

**Comment by jonblower** [comment]
Fair enough (although some systems would consider strings and URIs to be different primitives...)

Which reminds me, do we distinguish between integer and floating-point numbers in the dataType? Of course, categories are likely to be integers, but other things could be too (e.g. counts) and it might be useful for a client to know to expect only integers.

**Comment by neothemachine** [comment]
> Which reminds me, do we distinguish between integer and floating-point 
> numbers in the dataType? Of course, categories are likely to be 
> integers, but other things could be too (e.g. counts) and it might be 
> useful for a client to know to expect only integers.
>
Tricky question. JSON just supports "numbers", and when you load it into 
a JS object, you only get floating-point, with some hacks to make 
integers work somehow as well (don't ask). With typed arrays you can 
have real integers, but in my current implementations I only use those 
when the input already has them (CBOR), otherwise I create leave them as 
"number" arrays since I don't know the number type. So possibly it would 
have some advantages. I'll have to think about it.

**Comment by neothemachine** [comment]
Ok, closing this for now since string support is in. Please reopen if any of the other items from https://github.com/Reading-eScience-Centre/coveragejson/issues/28#issuecomment-151180842 should be considered.
